File: DSA/String/Code/SearchACharacter.py
# Search a Character in a String in Python
# Find position and count of a character

str1 = "Hello World"
target = 'l'

# Method 1: Using find() - returns index of first occurrence
print("Using find():")
pos = str1.find(target)
if pos != -1:
    print(f"Character '{target}' found at index: {pos}")
else:
    print(f"Character '{target}' not found")

# Method 2: Using rfind() - finds last occurrence
print("\nUsing rfind():")
lastPos = str1.rfind(target)
if lastPos != -1:
    print(f"Last occurrence of '{target}' at index: {lastPos}")

# Method 3: Using index() - similar to find() but raises exception
print("\nUsing index():")
try:
    pos = str1.index(target)
    print(f"Character '{target}' found at index: {pos}")
except ValueError:
    print(f"Character '{target}' not found")

# Method 4: Manual search with loop
print("\nManual search - all occurrences:")
count = 0
for i, char in enumerate(str1):
    if char == target:
        print(f"Found at index: {i}")
        count += 1
print(f"Total occurrences: {count}")

# Method 5: Using count() method
print("\nUsing count():")
total = str1.count(target)
print(f"Total occurrences of '{target}': {total}")

# Method 6: Finding all indices
print("\nFinding all indices:")
indices = [i for i, char in enumerate(str1) if char == target]
print(f"Indices: {indices}")

# Method 7: Case-insensitive search
print("\nCase-insensitive search for 'W':")
target2 = 'W'
lowerStr = str1.lower()
occurrences = lowerStr.count(target2.lower())
print(f"Case-insensitive count: {occurrences}")

# Method 8: Using string methods
print("\nUsing startswith and in operator:")
print(f"'{target}' in '{str1}': {target in str1}")

File: DSA/String/Code/SearchACharacter.cs
using System;
using System.Linq;

// Search a Character in a String in C#
// Find position and count of a character

class SearchACharacter {
    static void Main() {
        string str = "Hello World";
        char target = 'l';
        
        // Method 1: Using IndexOf() - finds first occurrence
        Console.WriteLine("Using IndexOf():");
        int pos = str.IndexOf(target);
        if (pos != -1) {
            Console.WriteLine("Character '" + target + "' found at index: " + pos);
        } else {
            Console.WriteLine("Character '" + target + "' not found");
        }
        
        // Method 2: Using LastIndexOf() - finds last occurrence
        Console.WriteLine("\nUsing LastIndexOf():");
        int lastPos = str.LastIndexOf(target);
        if (lastPos != -1) {
            Console.WriteLine("Last occurrence of '" + target + "' at index: " + lastPos);
        }
        
        // Method 3: Manual search with loop
        Console.WriteLine("\nManual search - all occurrences:");
        int count = 0;
        for (int i = 0; i < str.Length; i++) {
            if (str[i] == target) {
                Console.WriteLine("Found at index: " + i);
                count++;
            }
        }
        Console.WriteLine("Total occurrences: " + count);
        
        // Method 4: Finding all occurrences using loop
        Console.WriteLine("\nSearching from position 5:");
        int index = str.IndexOf(target, 5);
        while (index != -1) {
            Console.WriteLine("Found at index: " + index);
            index = str.IndexOf(target, index + 1);
        }
        
        // Method 5: Using LINQ Count
        Console.WriteLine("\nUsing LINQ Count():");
        int total = str.Count(c => c == target);
        Console.WriteLine("Total count of '" + target + "': " + total);
        
        // Method 6: Case-insensitive search
        Console.WriteLine("\nCase-insensitive search for 'W':");
        string lowerStr = str.ToLower();
        char lowerTarget = char.ToLower('W');
        int caseCount = lowerStr.Count(c => c == lowerTarget);
        Console.WriteLine("Case-insensitive count: " + caseCount);
        
        // Method 7: Finding all indices using LINQ
        Console.WriteLine("\nAll indices using LINQ:");
        var indices = str.Select((c, i) => new { Char = c, Index = i })
                          .Where(x => x.Char == target)
                          .Select(x => x.Index);
        Console.WriteLine("Indices: " + string.Join(", ", indices));
    }
}
